tools/smithy-rs-sync/src/main.rs

```diff
@@ -0,0 +1,409 @@
+        clean_out_existing_sdk(aws_sdk)
+            .context("couldn't clean out existing SDK from aws-sdk-rust")?;
+        copy_sdk(&build_artifacts, aws_sdk)?;
+        create_mirror_commit(&aws_sdk_repo, &commit)
+            .context("couldn't commit SDK changes to aws-sdk-rust")?;
+    }
+    println!("Successfully synced {} commit(s)", commit_revs.len());
+
+    // Get the last commit we synced so that we can set that for the next time this tool gets run
+    let last_synced_commit = commit_revs
+        .last()
+        .expect("can't be empty because we'd have early returned");
+    println!("Updating 'commit at last sync' to {}", last_synced_commit);
+
+    // Update the file containing the commit hash
+    set_last_synced_commit(&aws_sdk_repo, last_synced_commit).with_context(|| {
+        format!(
+            "couldn't write last synced commit hash ({}) to aws-sdk-rust/{}",
+            last_synced_commit, COMMIT_HASH_FILENAME,
+        )
+    })?;
+    // Commit the file containg the commit hash
+    commit_last_synced_commit_file(&aws_sdk_repo)
+        .context("couldn't commit the last synced commit hash file to aws-sdk-rust")?;
+
+    println!(
+        "Successfully synced all mirror commits to aws-sdk-rust/{}. Don't forget to push them",
+        branch
+    );
+
+    Ok(())
+}
+
+/// Starting from a given commit, walk the tree to its `HEAD` in order to build a list of commits that we'll
+/// need to sync. If you don't see the commits you're expecting, make sure the repo is up to date.
+/// This function doesn't include the `since_commit` in the list since that commit was synced last time
+/// this tool was run.
+fn commits_to_be_applied(smithy_rs_repo: &Repository, since_commit: &Oid) -> Result<Vec<Oid>> {
+    let rev_range = format!("{}..HEAD", since_commit);
+    println!("Checking for smithy-rs commits in range {}", rev_range);
+
+    let mut rev_walk = smithy_rs_repo.revwalk().context(here!())?;
+    rev_walk.push_range(&rev_range).context(here!())?;
+
+    let mut commit_revs = Vec::new();
+    for rev in rev_walk {
+        let rev = rev.context(here!())?;
+        commit_revs.push(rev);
+    }
+
+    // Order the revs from earliest to latest
+    commit_revs.reverse();
+
+    Ok(commit_revs)
+}
+
+/// Read the file from aws-sdk-rust that tracks the last smithy-rs commit it was synced with.
+/// Returns the hash of that commit.
+fn get_last_synced_commit(repo_path: &Path) -> Result<Oid> {
+    let path = repo_path.join(COMMIT_HASH_FILENAME);
+
+    let mut file = OpenOptions::new()
+        .read(true)
+        .open(&path)
+        .with_context(|| format!("couldn't open file at '{}'", path.to_string_lossy()))?;
+    // Commit hashes are 40 chars long
+    let mut commit_hash = String::with_capacity(40);
+    file.read_to_string(&mut commit_hash)
+        .with_context(|| format!("couldn't read file at '{}'", path.to_string_lossy()))?;
```

Comment:
is there a reason you didn't use `std::file::read_to_string(path)`?

Reply:
Not really, just a habit. Pedantically, I think this format is faster because we know the length of the string and can allocate it all at once.

Reply:
My intuition was incorrect. Under the hood, `std::fs::read_to_string` does smartly allocate